### ISSUE TITLE
Enhancement: Support multiple BASE_URL

### ISSUE THREAD

**Issue by timmc** [sandstorm-io/sandstorm#2771]
I'd like to have Sandstorm be configurable to support multiple base urls so that I can have a shared Sandstorm instances across several network protocols, or even just exposed on different domains.

The impetus here is that I'd like to expose a Sandstorm instance over [cjdns](https://github.com/cjdelisle/cjdns), which is a mesh network squatting on the fc00 IPv6 private address block; services running on cjdns have IP addresses that are not routable on the regular internet. I can't advertise the cjdns address of my Sandstorm instance in an AAAA record on my main domain, because that would break IPv6 address for regular internet users. That means I need a separate domain, and Sandstorm doesn't support that. (Even if I point my cjdns users at a proxy server which rewrites the *requests* to use the BASE_URL, the *responses* will point to the main BASE_URL as well.)

Cross-references:

- #434: Multiple host-name issue for Tor onion services
- #2011: Wildcard host and auto

(I'm admittedly new to cjdns; I might be wrong in my analysis.)

**Comment by ocdtrekkie** [comment]
So, I am thinking about this issue because I have internal resources, and I want to be sure I can access my Sandstorm server when the Internet is out, perhaps using my internal DNS (#1977 is a related need here). My Sandstorm server uses the same DNS server as my computers do, so I fear messing with the Sandcats DNS resolution would confuse my server and it's ability to renew certificates potentially. In this case, whatever wildcard host I connect with should continue to be used while I'm connected to the server.

I also can, more or less, I believe, migrate away from Sandcats in the near future. However, I do not want all of the existing URLs out there to documents on my server to break. So I want Sandcats to continue to renew certs, and I'd like it to redirect any grain IDs or share tokens attached to my Sandcats domain to my own, new domain (sandstorm-io/sandcats#162 is a related issue for this case).

Issue #2238 also discusses this issue a bit as well: When SANDCATS_BASE_DOMAIN is configured, we complain if the WILDCARD_HOST is not set for Sandcats. But we should continue to renew Sandcats certs either way, because we may want to redirect the user to the new wildcard host. (#1300 reminds me of another problem here... if people move away from Sandcats as their BASE_URL, we need a different config option to inform what the Sandcats domain is, since SANDCATS_BASE_DOMAIN is just set to sandcats.io to specify what server to talk to.)

Which is to say, I think there are probably a few things we can/should shore up here. However, now we'll have a whole new additional layer of issues: We might then also need to support multiple ACME accounts for TLS.

**Comment by deliciouslytyped** [comment]
Are these valid concerns here?: https://github.com/sandstorm-io/sandstorm/issues/2934#issuecomment-299667844

**Comment by xet7** [comment]
@deliciouslytyped 

No. Sandstorms works for me without Sandcats, when using CloudFlare DNS and SSL/TLS:

https://github.com/wekan/wekan/wiki/Sandstorm

@timmc 

Have you tried, does it work to configure AAAA with Caddy, that proxies to Sandstorm?

**Comment by timmc** [comment]
@ocdtrekkie I haven't messed around with cjdns or even IPv6 in many years, and I'm not really in a place where I can experiment with these things at the moment. Tor would still be interesting to me, though...

I'm not sure what you're saying about Sandcats. Are you saying that you've successfully run Sandstorm with multiple domains?

@deliciouslytyped That comment does seem relevant to me, yeah. :-/ Interestingly, those are both client-side issues. It seems to me like both of the things Kenton mentions could be mitigated by having Sandstorm be multi-domain aware and propagate that information appropriately. For example, on the OAuth side of things, it seems like Sandstorm should be able to answer differently depending on which domain name is being used for the OAuth flow. I might be missing something about identity unification, though -- OAuth is four protocols in a trenchcoat and I have a hard time keeping parts of it straight.

**Comment by xet7** [comment]
@timmc 

> I'm not sure what you're saying about Sandcats. Are you saying that you've successfully run Sandstorm with multiple domains?

I have only run one Sandstorm per domain. Previously with Sandcats, currently with CloudFlare. One Sandstorm can provide websites etc for many domains and subdomains.

**Comment by xet7** [comment]
@timmc 

It is possible to run multiple Sandstorm at same server. For example, one at localhost, others at KVM VMs. Then with Caddy configure what domains goes to localhost or KVM VM address. Domains can be login domains for each Sandstorm, or websites hosted at some Sandstorm.

**Comment by timmc** [comment]
Oh -- it sounds like you're talking about multiple Sandstorm servers at the same IP, each with a different domain. If so, that's different from single Sandstorm with multiple domains.

**Comment by xet7** [comment]
@timmc 

For single Meteor app at multiple domains, I have previously run many instances with different ROOT_URL settings, and same database. Maybe it could also work for Sandstorm, if multiple Sandstorm installs could be configured to use same MongoDB database?

https://github.com/wekan/wekan/blob/main/docker-compose.yml#L729-L753

**Comment by timmc** [comment]
I wouldn't assume that it's safe to do that. Maybe Blackrock could do it, but single-host Sandstorm may have a lot of assumptions baked into it about singleton access and mutual exclusion.

**Comment by ocdtrekkie** [comment]
I still really want to have a way to do this for "migrating off Sandcats without breaking existing links" purposes... but I do split-brain my DNS so that my Sandstorm works when the Internet is down.

Currently, you can have both a HTTPS_PORT and PORT set, and the http links will be automatically redirected to https links. But there's no easy way to redirect the BASE_URL, even for avoiding-breaking-links purposes.
